spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala

```diff
@@ -1024,7 +1024,25 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       }
     }
   }
+  test("unhex") {
+    val table = "unhex_table"
+    withTable(table) {
+      sql(s"create table $table(col string) using parquet")
+
+      sql(s"""INSERT INTO $table VALUES
+        |('537061726B2053514C'),
+        |('737472696E67'),
+        |('\0'),
+        |(''),
+        |('###'),
+        |('G123'),
+        |('hello'),
+        |('A1B'),
+        |('0A1B')""".stripMargin)
 
+      checkSparkAnswerAndOperator(s"SELECT unhex(col) FROM $table")
```

Comment:
It may be worth adding an `ORDER BY` clause here to ensure that the test is deterministic.

```suggestion
      checkSparkAnswerAndOperator(s"SELECT unhex(col) FROM $table ORDER BY col")
```

Reply:
I wonder if the Spark 3.2 failure is at all related to ordering?


```
   [WrappedArray(10, 27)]                                  [WrappedArray(10, 27)]
  ![WrappedArray(115, 116, 114, 105, 110, 103)]            [WrappedArray(10, 27)]
  ![WrappedArray(27, 0)]                                   [WrappedArray(115, 116, 114, 105, 110, 103)]
  ```

Reply:
Thanks, yeah, I looked at that bit. I think the test results are sorted, but the Spark case (left column) 27 sorts after 115.

Unfortunately, I think it may be legitimately different behavior between the 3.2 and the more recent versions.
<img width="995" alt="image" src="https://github.com/apache/datafusion-comet/assets/421839/36c59275-83b9-4099-9692-9913313f85c7"> .

Reply:
Looking at the spark repo a bit more, it looks like there was a bug in the [3.2 implementation](https://github.com/apache/spark/commit/276abe3c3311300cffc6570b66f3977ea8172ff0) that was fixed in subsequent versions.

How should I handle this? Like should I write a native implementation of the erroneous code in order to be faithful to the spark, or have it be correct if different?

Reply:
Actually adding the `ORDER BY` causes the test to fail because it no longer runs fully native, so ignore that suggestion.

Also, I do see code differences in Spark between 3.2 and 3.4 in the unhex algorithm. 3.2 does not have the `oddShift` variable. `oddShift` was added in https://github.com/apache/spark/commit/276abe3c3311300cffc6570b66f3977ea8172ff0

I guess the options are:

- mark `unhex` as incompat just for 3.2 and skip the test for 3.2 (probably the easiest path)
- implement per-spark-version logic in Rust for unhex

Let me know what you think or if you have any questions on this. This is a good example of the challenge of supporting multiple Spark versions 😓

Reply:
Thanks for looking, I think we found the same commit w.r.t. 3.2's differences. 

I'll just go the easy route since matching 3.2 would be implementing buggy code and 3.2 is the oldest supported version (so presumably it'll be dropped at some point). I'll use this as a chance to poke around the repo, but may ask in discord if I get blocked on marking it incompat for 3.2 and skipping the tests for that version.

Reply:
Made an update in https://github.com/apache/datafusion-comet/pull/342/commits/36baf8e87dd2689baf8c45961c7ed34a97f1f493 that seems to work (https://github.com/tshauck/arrow-datafusion-comet/actions/runs/8994978806). Happy to get any feedback about the approach taken.